core/src/org/pentaho/platform/engine/core/system/PentahoSystem.java

```diff
@@ -549,72 +532,47 @@ public static int getInitializedStatus() {
-      rtn.add( Messages.getInstance().getString( "PentahoSystem.USER_INITIALIZATION_SYSTEM_LISTENERS_FAILED" ) ); //$NON-NLS-1$
-      List l = PentahoSystem.getAdditionalInitializationFailureMessages( PentahoSystem.SYSTEM_LISTENERS_FAILED );
-      if ( l != null ) {
-        rtn.addAll( l );
-      }
-    }
-    if ( PentahoSystem.hasFailed( PentahoSystem.SYSTEM_PENTAHOXML_FAILED ) ) {
-      rtn.add( Messages.getInstance().getString( "PentahoSystem.USER_INITIALIZATION_SYSTEM_PENTAHOXML_FAILED" ) ); //$NON-NLS-1$
-      List l = PentahoSystem.getAdditionalInitializationFailureMessages( PentahoSystem.SYSTEM_PENTAHOXML_FAILED );
-      if ( l != null ) {
-        rtn.addAll( l );
-      }
-    }
-    if ( PentahoSystem.hasFailed( PentahoSystem.SYSTEM_OTHER_FAILED ) ) {
-      rtn.add( Messages.getInstance().getString( "PentahoSystem.USER_INITIALIZATION_SYSTEM_OTHER_FAILED" ) ); //$NON-NLS-1$
-      List l = PentahoSystem.getAdditionalInitializationFailureMessages( PentahoSystem.SYSTEM_OTHER_FAILED );
-      if ( l != null ) {
-        rtn.addAll( l );
+  private static void getFailureMessages( List<String> errors, final Integer failureBit, String messageKey,
+      Object... params ) {
+    if ( PentahoSystem.hasFailed( failureBit ) ) {
+      errors.add( Messages.getInstance().getString( "PentahoSystem." + messageKey, params ) );//$NON-NLS-1$
+      List<String> l = PentahoSystem.getAdditionalInitializationFailureMessages( failureBit );
+      if ( CollectionUtils.isNotEmpty( l ) ) {
+        errors.addAll( l );
       }
     }
+  }
+
+  public static List<String> getInitializationFailureMessages() {
+    List<String> rtn = new ArrayList<String>();
+
+    getFailureMessages( rtn, PentahoSystem.SYSTEM_SETTINGS_FAILED,
+        "USER_INITIALIZATION_SYSTEM_SETTINGS_FAILED", PathBasedSystemSettings.SYSTEM_CFG_PATH_KEY );//$NON-NLS-1$
+
+    getFailureMessages( rtn, PentahoSystem.SYSTEM_PUBLISHERS_FAILED, "USER_INITIALIZATION_SYSTEM_PUBLISHERS_FAILED" );//$NON-NLS-1$
+
+    getFailureMessages( rtn, PentahoSystem.SYSTEM_OBJECTS_FAILED, "USER_INITIALIZATION_SYSTEM_OBJECTS_FAILED" );//$NON-NLS-1$
+
+    getFailureMessages( rtn, PentahoSystem.SYSTEM_AUDIT_FAILED, "USER_INITIALIZATION_SYSTEM_AUDIT_FAILED" );//$NON-NLS-1$
+
+    getFailureMessages( rtn, PentahoSystem.SYSTEM_LISTENERS_FAILED, "USER_INITIALIZATION_SYSTEM_LISTENERS_FAILED" );//$NON-NLS-1$
+
+    getFailureMessages( rtn, PentahoSystem.SYSTEM_PENTAHOXML_FAILED, "USER_INITIALIZATION_SYSTEM_PENTAHOXML_FAILED" );//$NON-NLS-1$
+
+    getFailureMessages( rtn, PentahoSystem.SYSTEM_OTHER_FAILED, "USER_INITIALIZATION_SYSTEM_OTHER_FAILED" );//$NON-NLS-1$
+
     return rtn;
   }
 
-  public static void addInitializationFailureMessage( final int failureBit, final String message ) {
-    Integer i = new Integer( failureBit );
-    List l = (List) PentahoSystem.initializationFailureDetailsMap.get( i );
+  public static void addInitializationFailureMessage( final Integer failureBit, final String message ) {
```

Comment:
I agree with you and I will revert that, because my code can write null into the map instead nullpointer exception.